warp/src/tesseract/mod.rs

```diff
@@ -478,29 +499,96 @@ impl Tesseract {
+#[cfg_attr(target_arch = "wasm32", wasm_bindgen)]
+impl Tesseract {
+    #[wasm_bindgen(js_name = set)]
+    pub fn set_wasm(&self, key: &str, value: &str) -> std::result::Result<(), JsError> {
+        self.inner.set(key, value).map_err(|e| e.into())
+    }
+
+    #[wasm_bindgen(js_name = retrieve)]
+    pub fn retrieve_wasm(&self, key: &str) -> std::result::Result<String, JsError> {
+        self.inner.retrieve(key).map_err(|e| e.into())
+    }
+
+    #[wasm_bindgen(js_name = update_unlock)]
+    pub fn update_unlock_wasm(
+        &self,
+        old_passphrase: &[u8],
+        new_passphrase: &[u8],
+    ) -> std::result::Result<(), JsError> {
+        self.inner
+            .update_unlock(old_passphrase, new_passphrase)
+            .map_err(|e| e.into())
+    }
+
+    #[wasm_bindgen(js_name = delete)]
+    pub fn delete_wasm(&self, key: &str) -> std::result::Result<(), JsError> {
+        self.inner.delete(key).map_err(|e| e.into())
+    }
+
+    #[wasm_bindgen(js_name = unlock)]
+    pub fn unlock_wasm(&self, passphrase: &[u8]) -> std::result::Result<(), JsError> {
+        self.inner.unlock(passphrase).map_err(|e| e.into())
+    }
+
+    #[wasm_bindgen(js_name = save)]
+    pub fn save_wasm(&self) -> std::result::Result<(), JsError> {
+        self.inner.save().map_err(|e| e.into())
+    }
+
+    #[wasm_bindgen(js_name = subscribe)]
+    pub fn subscribe_wasm(&self) -> AsyncIterator {
+        Into::<JsValue>::into(Subscription {
+            inner: self.inner.subscribe(),
+        })
+        .into()
+    }
+
+    /// Used to load contents from local storage
+    pub fn load_from_storage(&self) -> std::result::Result<(), JsError> {
+        self.inner.load_from_storage().map_err(|e| e.into())
     }
 }
 
-impl Tesseract {
-    pub fn subscribe(&self) -> BoxStream<'static, TesseractEvent> {
-        self.inner.subscribe()
+/// Wraps BoxStream<'static, TesseractEvent> into a js compatible struct
+#[cfg(target_arch = "wasm32")]
+#[wasm_bindgen]
+pub struct Subscription {
+    inner: BoxStream<'static, TesseractEvent>,
+}
+
+/// Provides the next() function expected by js async iterator
+#[cfg(target_arch = "wasm32")]
+#[wasm_bindgen]
+impl Subscription {
+    pub async fn next(&mut self) -> std::result::Result<Promise, JsError> {
+        let next = self.inner.next().await;
+        match next {
+            Some(value) => Ok(Promise::resolve(
+                &TesseractEventPromiseResult::new(value).into(),
+            )),
+            None => std::result::Result::Err(JsError::new("returned None")),
+        }
     }
 }
 
+/// Wraps in the TesseractEvent promise result in the js object expected by js async iterator
 #[cfg(target_arch = "wasm32")]
-impl Tesseract {
-    /// Used to load contents from local storage
-    pub fn load_from_storage(&self) -> Result<()> {
-        self.inner.load_from_storage()
+#[wasm_bindgen]
+struct TesseractEventPromiseResult {
+    pub value: TesseractEvent,
+    pub done: bool,
```

Comment:
This does not look used or is this used when exported?

Reply:
This is used by asynciterator. The Promise returned by `next()` must be a js `object` with the `value` and `done` fields.